frappe/core/doctype/version/version.py

```diff
@@ -2,14 +2,23 @@
 # License: MIT. See LICENSE
 
 import json
+from typing import Optional
 
 import frappe
 from frappe.model import no_value_fields, table_fields
 from frappe.model.document import Document
 
 
 class Version(Document):
-	def set_diff(self, old, new):
+	def update_version_info(self, old: Optional[Document], new: Document) -> bool:
```

Comment:
I think this should include subclasses of `Document` too, right?

```suggestion
	def update_version_info(self, old: Optional[Type[Document]], new: Type[Document]) -> bool:
```

Reply:
Typing by default allows subclasses.

`Type[class]` is different, it means class object instead of instance objects. 

Document is fine here IMO.

Reply:
https://docs.python.org/3/library/typing.html#typing.Type

Reply:
`Type[class]` actually was also meant for subclassing (check the example in your link), but either way I just saw the note that it's deprecated, so we're good 😅

Reply:
> Type[class] actually was also meant for subclassing (check the example in your link), 

It's not? Can you share which specific examples makes you think that? 

`c = type(a)` Here c is the "type" int and not instances of it.

Reply:
I was referring to this example:

```
Note that Type[C] is covariant:

class User: ...
class BasicUser(User): ...
class ProUser(User): ...
class TeamUser(User): ...

# Accepts User, BasicUser, ProUser, TeamUser, ...
def make_new_user(user_class: Type[User]) -> User:
    # ...
    return user_class()
```

I realize that it's just a property of the `Type`, but it does allow subclasses via covariance

Reply:
It's accepting different subclasses of User which is not same as objects of those classes. Relook at the example, it's using the passed class to instantiate a new object from it. 

---

In frappe terms:
```
Customer = get_controller("Customer") # -> Type[Document]
customer_doc = frappe.get_doc("Customer, "Google") # -> Document
```


Here version takes in instances of the documents and compares them, and has nothing to do with their types of them. 


Though more specific typing would be 


```python
D = TypeVar("D", bound="Document")
def update_version_info(self, old: D, new: D) -> bool:
    ...
```


I avoid going much in depth with typing because we aren't type checking anyway, so specific types are mostly waste right now :p

Reply:
Oh, I understand now. Thanks for the explanation!